## Analiza Danych Sprzedażowych

## Import

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


## Wczytanie danych

In [2]:
df = pd.read_csv('./sales_data_sample.csv', encoding='latin1')

df.head()

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0        10107               30      95.70                2  2871.00   
1        10121               34      81.35                5  2765.90   
2        10134               41      94.74                2  3884.34   
3        10145               45      83.26                6  3746.70   
4        10159               49     100.00               14  5205.27   

         ORDERDATE   STATUS  QTR_ID  MONTH_ID  YEAR_ID  ...  \
0   2/24/2003 0:00  Shipped       1         2     2003  ...   
1    5/7/2003 0:00  Shipped       2         5     2003  ...   
2    7/1/2003 0:00  Shipped       3         7     2003  ...   
3   8/25/2003 0:00  Shipped       3         8     2003  ...   
4  10/10/2003 0:00  Shipped       4        10     2003  ...   

                    ADDRESSLINE1  ADDRESSLINE2           CITY STATE  \
0        897 Long Airport Avenue           NaN            NYC    NY   
1             59 rue de l'Abbaye           NaN          Reims   NaN   
2  27 rue du Colonel Pierre Avia           NaN          Paris   NaN   
3             78934 Hillside Dr.           NaN       Pasadena    CA   
4                7734 Strong St.           NaN  San Francisco    CA   

  POSTALCODE COUNTRY TERRITORY CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0      10022     USA       NaN              Yu             Kwai    Small  
1      51100  France      EMEA         Henriot             Paul    Small  
2      75508  France      EMEA        Da Cunha           Daniel   Medium  
3      90003     USA       NaN           Young            Julie   Medium  
4        NaN     USA       NaN           Brown            Julie   Medium  

[5 rows x 25 columns]

## Eksploracja danych

In [3]:
df.columns = [col.lower() for col in df.columns]
df

ordernumber  quantityordered  priceeach  orderlinenumber    sales  \
0           10107               30      95.70                2  2871.00   
1           10121               34      81.35                5  2765.90   
2           10134               41      94.74                2  3884.34   
3           10145               45      83.26                6  3746.70   
4           10159               49     100.00               14  5205.27   
...           ...              ...        ...              ...      ...   
2818        10350               20     100.00               15  2244.40   
2819        10373               29     100.00                1  3978.51   
2820        10386               43     100.00                4  5417.57   
2821        10397               34      62.24                1  2116.16   
2822        10414               47      65.52                9  3079.44   

            orderdate    status  qtr_id  month_id  year_id  ...  \
0      2/24/2003 0:00   Shipped       1         2     2003  ...   
1       5/7/2003 0:00   Shipped       2         5     2003  ...   
2       7/1/2003 0:00   Shipped       3         7     2003  ...   
3      8/25/2003 0:00   Shipped       3         8     2003  ...   
4     10/10/2003 0:00   Shipped       4        10     2003  ...   
...               ...       ...     ...       ...      ...  ...   
2818   12/2/2004 0:00   Shipped       4        12     2004  ...   
2819   1/31/2005 0:00   Shipped       1         1     2005  ...   
2820    3/1/2005 0:00  Resolved       1         3     2005  ...   
2821   3/28/2005 0:00   Shipped       1         3     2005  ...   
2822    5/6/2005 0:00   On Hold       2         5     2005  ...   

                       addressline1  addressline2           city state  \
0           897 Long Airport Avenue           NaN            NYC    NY   
1                59 rue de l'Abbaye           NaN          Reims   NaN   
2     27 rue du Colonel Pierre Avia           NaN          Paris   NaN   
3                78934 Hillside Dr.           NaN       Pasadena    CA   
4                   7734 Strong St.           NaN  San Francisco    CA   
...                             ...           ...            ...   ...   
2818             C/ Moralzarzal, 86           NaN         Madrid   NaN   
2819                    Torikatu 38           NaN           Oulu   NaN   
2820             C/ Moralzarzal, 86           NaN         Madrid   NaN   
2821          1 rue Alsace-Lorraine           NaN       Toulouse   NaN   
2822             8616 Spinnaker Dr.           NaN         Boston    MA   

     postalcode  country territory contactlastname contactfirstname dealsize  
0         10022      USA       NaN              Yu             Kwai    Small  
1         51100   France      EMEA         Henriot             Paul    Small  
2         75508   France      EMEA        Da Cunha           Daniel   Medium  
3         90003      USA       NaN           Young            Julie   Medium  
4           NaN      USA       NaN           Brown            Julie   Medium  
...         ...      ...       ...             ...              ...      ...  
2818      28034    Spain      EMEA          Freyre            Diego    Small  
2819      90110  Finland      EMEA       Koskitalo           Pirkko   Medium  
2820      28034    Spain      EMEA          Freyre            Diego   Medium  
2821      31000   France      EMEA          Roulet          Annette    Small  
2822      51003      USA       NaN         Yoshido             Juri   Medium  

[2823 rows x 25 columns]

In [4]:
df.describe().T

count          mean          std       min       25%  \
ordernumber      2823.0  10258.725115    92.085478  10100.00  10180.00   
quantityordered  2823.0     35.092809     9.741443      6.00     27.00   
priceeach        2823.0     83.658544    20.174277     26.88     68.86   
orderlinenumber  2823.0      6.466171     4.225841      1.00      3.00   
sales            2823.0   3553.889072  1841.865106    482.13   2203.43   
qtr_id           2823.0      2.717676     1.203878      1.00      2.00   
month_id         2823.0      7.092455     3.656633      1.00      4.00   
year_id          2823.0   2003.815090     0.699670   2003.00   2003.00   
msrp             2823.0    100.715551    40.187912     33.00     68.00   

                     50%      75%      max  
ordernumber      10262.0  10333.5  10425.0  
quantityordered     35.0     43.0     97.0  
priceeach           95.7    100.0    100.0  
orderlinenumber      6.0      9.0     18.0  
sales             3184.8   4508.0  14082.8  
qtr_id               3.0      4.0      4.0  
month_id             8.0     11.0     12.0  
year_id           2004.0   2004.0   2005.0  
msrp                99.0    124.0    214.0

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2823 entries, 0 to 2822
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ordernumber       2823 non-null   int64  
 1   quantityordered   2823 non-null   int64  
 2   priceeach         2823 non-null   float64
 3   orderlinenumber   2823 non-null   int64  
 4   sales             2823 non-null   float64
 5   orderdate         2823 non-null   object 
 6   status            2823 non-null   object 
 7   qtr_id            2823 non-null   int64  
 8   month_id          2823 non-null   int64  
 9   year_id           2823 non-null   int64  
 10  productline       2823 non-null   object 
 11  msrp              2823 non-null   int64  
 12  productcode       2823 non-null   object 
 13  customername      2823 non-null   object 
 14  phone             2823 non-null   object 
 15  addressline1      2823 non-null   object 
 16  addressline2      302 non-null    object 


In [6]:
df.columns

Index(['ordernumber', 'quantityordered', 'priceeach', 'orderlinenumber',
       'sales', 'orderdate', 'status', 'qtr_id', 'month_id', 'year_id',
       'productline', 'msrp', 'productcode', 'customername', 'phone',
       'addressline1', 'addressline2', 'city', 'state', 'postalcode',
       'country', 'territory', 'contactlastname', 'contactfirstname',
       'dealsize'],
      dtype='object')

In [7]:
df[['year_id']].isnull().sum()

year_id    0
dtype: int64

In [8]:
df[['month_id']].isnull().sum()

month_id    0
dtype: int64

In [9]:
df.dtypes

ordernumber           int64
quantityordered       int64
priceeach           float64
orderlinenumber       int64
sales               float64
orderdate            object
status               object
qtr_id                int64
month_id              int64
year_id               int64
productline          object
msrp                  int64
productcode          object
customername         object
phone                object
addressline1         object
addressline2         object
city                 object
state                object
postalcode           object
country              object
territory            object
contactlastname      object
contactfirstname     object
dealsize             object
dtype: object

## Konwersja kolumny 'orderdate' oraz dodanie kolumn 'order_day' oraz 'order_month'

In [10]:
df['orderdate'] = pd.to_datetime(df['orderdate'])

In [11]:
df.head()

ordernumber  quantityordered  priceeach  orderlinenumber    sales  \
0        10107               30      95.70                2  2871.00   
1        10121               34      81.35                5  2765.90   
2        10134               41      94.74                2  3884.34   
3        10145               45      83.26                6  3746.70   
4        10159               49     100.00               14  5205.27   

   orderdate   status  qtr_id  month_id  year_id  ...  \
0 2003-02-24  Shipped       1         2     2003  ...   
1 2003-05-07  Shipped       2         5     2003  ...   
2 2003-07-01  Shipped       3         7     2003  ...   
3 2003-08-25  Shipped       3         8     2003  ...   
4 2003-10-10  Shipped       4        10     2003  ...   

                    addressline1  addressline2           city state  \
0        897 Long Airport Avenue           NaN            NYC    NY   
1             59 rue de l'Abbaye           NaN          Reims   NaN   
2  27 rue du Colonel Pierre Avia           NaN          Paris   NaN   
3             78934 Hillside Dr.           NaN       Pasadena    CA   
4                7734 Strong St.           NaN  San Francisco    CA   

  postalcode country territory contactlastname contactfirstname dealsize  
0      10022     USA       NaN              Yu             Kwai    Small  
1      51100  France      EMEA         Henriot             Paul    Small  
2      75508  France      EMEA        Da Cunha           Daniel   Medium  
3      90003     USA       NaN           Young            Julie   Medium  
4        NaN     USA       NaN           Brown            Julie   Medium  

[5 rows x 25 columns]

In [12]:
df['order_day'] = df['orderdate'].dt.day_of_week

In [13]:
df['order_month'] = df['orderdate'].dt.month

In [14]:
df.head()

ordernumber  quantityordered  priceeach  orderlinenumber    sales  \
0        10107               30      95.70                2  2871.00   
1        10121               34      81.35                5  2765.90   
2        10134               41      94.74                2  3884.34   
3        10145               45      83.26                6  3746.70   
4        10159               49     100.00               14  5205.27   

   orderdate   status  qtr_id  month_id  year_id  ...           city  state  \
0 2003-02-24  Shipped       1         2     2003  ...            NYC     NY   
1 2003-05-07  Shipped       2         5     2003  ...          Reims    NaN   
2 2003-07-01  Shipped       3         7     2003  ...          Paris    NaN   
3 2003-08-25  Shipped       3         8     2003  ...       Pasadena     CA   
4 2003-10-10  Shipped       4        10     2003  ...  San Francisco     CA   

  postalcode country territory contactlastname contactfirstname dealsize  \
0      10022     USA       NaN              Yu             Kwai    Small   
1      51100  France      EMEA         Henriot             Paul    Small   
2      75508  France      EMEA        Da Cunha           Daniel   Medium   
3      90003     USA       NaN           Young            Julie   Medium   
4        NaN     USA       NaN           Brown            Julie   Medium   

  order_day order_month  
0         0           2  
1         2           5  
2         1           7  
3         0           8  
4         4          10  

[5 rows x 27 columns]

## Zadania: 
 - Jak wygląda miesięczna sprzedaż (ilościowa i wartościowa)?

 - Jaki był średni przychód na jedno zamówienie?

 - Jaki był średni przychód na jedną zamówioną sztukę?

 - Który miesiąc był najlepszy pod względem sprzedaży (ilościowo i wartościowo)?

 - Jak wygląda sprzedaż na wykresach: miesięczna ilościowa, wartościowa i porównanie średnich przychodów?

In [15]:
sales_per_month = df[['order_month', 'quantityordered']]
sales_per_month

order_month  quantityordered
0               2               30
1               5               34
2               7               41
3               8               45
4              10               49
...           ...              ...
2818           12               20
2819            1               29
2820            3               43
2821            3               34
2822            5               47

[2823 rows x 2 columns]

In [16]:
sales_per_month = sales_per_month.sort_values('order_month')
sales_per_month= sales_per_month.groupby('order_month').sum()

In [17]:
print(sales_per_month.columns)


Index(['quantityordered'], dtype='object')


In [18]:
sales_per_month = sales_per_month.reset_index()
sales_per_month = sales_per_month.rename(columns={'quantityordered': 'order_values'})
sales_per_month

order_month  order_values
0             1          7997
1             2          7903
2             3          7585
3             4          6704
4             5          8992
5             6          4620
6             7          4899
7             8          6538
8             9          5681
9            10         10998
10           11         20857
11           12          6293

In [19]:
sales_per_month['order_values'].max()

np.int64(20857)

In [20]:
best_month = sales_per_month['order_values'] == sales_per_month['order_values'].max()
sales_per_month[best_month]

order_month  order_values
10           11         20857

Najlepszy miesiąc to Listopad z poziomem sprzedaży na poziomie 20857 zamówień.

In [21]:
df.head()

ordernumber  quantityordered  priceeach  orderlinenumber    sales  \
0        10107               30      95.70                2  2871.00   
1        10121               34      81.35                5  2765.90   
2        10134               41      94.74                2  3884.34   
3        10145               45      83.26                6  3746.70   
4        10159               49     100.00               14  5205.27   

   orderdate   status  qtr_id  month_id  year_id  ...           city  state  \
0 2003-02-24  Shipped       1         2     2003  ...            NYC     NY   
1 2003-05-07  Shipped       2         5     2003  ...          Reims    NaN   
2 2003-07-01  Shipped       3         7     2003  ...          Paris    NaN   
3 2003-08-25  Shipped       3         8     2003  ...       Pasadena     CA   
4 2003-10-10  Shipped       4        10     2003  ...  San Francisco     CA   

  postalcode country territory contactlastname contactfirstname dealsize  \
0      10022     USA       NaN              Yu             Kwai    Small   
1      51100  France      EMEA         Henriot             Paul    Small   
2      75508  France      EMEA        Da Cunha           Daniel   Medium   
3      90003     USA       NaN           Young            Julie   Medium   
4        NaN     USA       NaN           Brown            Julie   Medium   

  order_day order_month  
0         0           2  
1         2           5  
2         1           7  
3         0           8  
4         4          10  

[5 rows x 27 columns]

In [22]:
sum_ordered = df['quantityordered'].sum()
sum_sales = df['sales'].sum()

In [23]:
avg = sum_sales/ sum_ordered 
avg

np.float64(101.27114831376744)

Średni przychód na jedno zamówienie wynosi 101,27$

In [24]:
df.columns

Index(['ordernumber', 'quantityordered', 'priceeach', 'orderlinenumber',
       'sales', 'orderdate', 'status', 'qtr_id', 'month_id', 'year_id',
       'productline', 'msrp', 'productcode', 'customername', 'phone',
       'addressline1', 'addressline2', 'city', 'state', 'postalcode',
       'country', 'territory', 'contactlastname', 'contactfirstname',
       'dealsize', 'order_day', 'order_month'],
      dtype='object')

 - Jaki był średni przychód na jedną zamówioną sztukę?